java-checks/src/main/java/org/sonar/java/checks/regex/UnusedGroupNamesCheck.java

```diff
@@ -47,13 +48,27 @@
 
   private static final String JAVA_UTIL_REGEX_MATCHER = "java.util.regex.Matcher";
 
-  private static final MethodMatchers MATCHER_GROUP = MethodMatchers.create()
-    .ofTypes(JAVA_UTIL_REGEX_MATCHER)
-    .names("group")
-    // covers both 'group(String)' and 'group(int)'
-    .addParametersMatcher(JAVA_LANG_STRING)
-    .addParametersMatcher("int")
-    .build();
+  private static final Pattern GROUP_NUMBER_REPLACEMENT_REGEX = Pattern.compile("(?<!\\\\)\\$(?<number>\\d++)");
+  private static final Pattern GROUP_NAME_REPLACEMENT_REGEX = Pattern.compile("(?<!\\\\)\\$\\{(?<name>[A-Za-z][0-9A-Za-z]*+)\\}");
+
+  private static final MethodMatchers MATCHER_GROUP = MethodMatchers.or(
+    MethodMatchers.create()
+      .ofTypes(JAVA_UTIL_REGEX_MATCHER)
+      .names("group")
+      // covers both 'group(String)' and 'group(int)'
+      .addParametersMatcher(JAVA_LANG_STRING)
+      .addParametersMatcher("int")
+      .build(),
+    MethodMatchers.create()
+      .ofTypes(JAVA_UTIL_REGEX_MATCHER)
+      .names("appendReplacement")
+      .addParametersMatcher(MethodMatchers.ANY, MethodMatchers.ANY)
+      .build(),
+    MethodMatchers.create()
+      .ofTypes(JAVA_UTIL_REGEX_MATCHER)
+      .names("replaceAll", "replaceFirst")
+      .addParametersMatcher(MethodMatchers.ANY)
```

Comment:
I was wondering: is there any specific reason to chose `ANY` instead of `java.lang.String`?

Reply:
Because java 9 introduces more method signatures to the `Matcher`:
```
replaceAll(Function<MatchResult, String> replacer)
replaceFirst(Function<MatchResult, String> replacer)
```
And finding all replaceFirst, replaceAll and appendReplacement methods is important to not have false-positive.
Not being able to convert the argument to a string constant is not a problem in the rule implementation.
And I didn't add a unit test about such false-positive because we don't support java 9 in our compiled unit tests. And the rule requires a project that compiles to have the semantic about replaceFirst, replaceAll.

Reply:
Make sense, thanks for the explanation. 👍